### PR TITLE
mtl: drv: Power off flow, HPSRAM memory save/restore to an external buffer

### PR DESCRIPTION
**Pull request by tmleman**
This PR adds a power flows to Intel MTL TLB driver
It allows the driver to save the whole Translation Table
as well as contents of memory banks to a designated address
It is designed for memory content retention in an persistent
storage during low power state

### CODE REVIEW

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 1]
```diff
@@ -693,4 +698,162 @@ static int sys_mm_drv_mm_init(const struct device *dev)
 	return 0;
 }
 
-SYS_INIT(sys_mm_drv_mm_init, POST_KERNEL, 0);
+#ifdef CONFIG_PM_DEVICE
+static void sys_mm_save_context(void *storage_buffer)
```
Comment (dcpleung) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:702]: Please do not use the `sys_mm_` namespace for these newly added functions as these are not official APIs.

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi [hunk 2]
```diff
@@ -125,6 +125,7 @@
 		tlb: tlb@17e000 {
 			compatible = "intel,adsp-tlb";
 			reg = <0x17e000 0x1000>;
+			label = "HPSRAM_TLB";
```
Comment (dcpleung) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:128]: Actually, the main branch is in the process of removing usage of `label` properties in device tree. So please change the code so this does not require this.

**Review by dcpleung — CHANGES_REQUESTED** [review]

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c (hunk 1 above)
Comment (tmleman) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c]: @dcpleung please give me some suggestions on function names if the new ones turn out to be bad.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 3]
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
...
+}
+
+static int adsp_mm_drv_pm_callback(const struct device *dev,
+				  enum pm_device_action action)
+{
+	(void) dev;
+	switch (action) {
+	case PM_DEVICE_ACTION_FORCE_SUSPEND:
+		if (context_save_buffer != NULL) {
+			sys_mm_save_context(context_save_buffer);
```
Comment (dcpleung) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:820]: Need to use new name.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 4]
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
...
+		}
+		break;
+
+	case PM_DEVICE_ACTION_RESUME:
+		if (context_save_buffer != NULL) {
+			sys_mm_restore_context(context_save_buffer);
```
Comment (dcpleung) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:826]: Need to use new name.

**Review by dcpleung — DISMISSED** [review]

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c (hunk 1 above)
Comment (dcpleung) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c]: New names look good.

**Review by aborisovich — DISMISSED** [review]

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 5]
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
...
 		return -EFAULT;
 	}
 
 	/*
 	 * Unmap all unused physical pages from the entire
 	 * virtual address space to save power
 	 */
-	size_t unused_size = CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE -
-			     unused_l2_start_aligned;
+	size_t unused_size = L2_SRAM_BASE + L2_SRAM_SIZE - UNUSED_L2_START_ALIGNED;
```
Comment (jxstelter) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:697]: Here you reverted change for virtual memory Fix. Virtual memory space is much higher than physical one.

**Review by softwarecki — DISMISSED** [review]

**Review by jxstelter — CHANGES_REQUESTED** [review]
This patch reverts fix:
SHA-1: afb156a1eb217c5b7bdedddb75752c697cae381a

* soc: ace_v1x: unmap entire unused virtual space

This patch corrects the initial unmapping of unused l2 memory
to unmap until the end of virtual address space instead
of unmapping until the end of l2 physical memory.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 6]
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
...
+	location += TLB_SIZE;
+
+	/* save context of all the pages that had been mapped */
+	/* FIXME - currently saving the memory above the unused marker only
+	 * according to needs there may be more memory saved, even the whole HPSRAM (start_page=0)
+	 */
+	uint32_t start_page = (UNUSED_L2_START_ALIGNED -  POINTER_TO_UINT(L2_SRAM_BASE))
+					/ CONFIG_MM_DRV_PAGE_SIZE;
+
+	for (page_idx = start_page; page_idx < L2_SRAM_PAGES_NUM; page_idx++) {
```
Comment (jxstelter) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:732]: This loop will go only in physical memory range while TLB table is for virtual range. Now it will loose entries for virtual addresses higher than L2_SRAM_BASE + (L2_SRAM_PAGES_NUM * CONFIG_MM_DRV_PAGE_SIZE). 
It should go from CONFIG_KERNEL_VM_BASE to (CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE)

Comment (marcinszkudlinski) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:732]: that's not true.
see line 743 - there's a remap of used physical memory pages 1:1 to virtual space, so I don't care about how the physical pages used to be mapped to a virtual space. Its enough to store all physical memory - mapped 1:1 to virtual!

This is much faster than checking all virtual mapping and taking care for regions mapped to several addresses etc.

**Review by softwarecki — DISMISSED** [review]

**Review by jxstelter — DISMISSED** [review]

File: soc/xtensa/intel_adsp/ace_v1x/power.c [hunk 7]
```diff
@@ -43,7 +46,18 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		if (cpu == 0) {
 			/* FIXME: this value should come from MM */
 			uint32_t hpsram_mask[1] = { 0x3FFFFF };
+#if CONFIG_PM_DEVICE
+			const struct device *dev = device_get_binding("HPSRAM_TLB");
```
Comment (dcpleung) [inline soc/xtensa/intel_adsp/ace_v1x/power.c:50]: We are in the process of getting rid of labels in DTS. So it would be great if this is done using `DEVICE_DT_GET(DT_NODELABEL(node_label))`. Also, there is no `HPSRAM_TLB` in the tree right now. So if we enable `CONFIG_PM_DEVICE`, this won't build at all.

**Review by andyross — COMMENTED** [review]
Isn't this a layering violation?   The mm_drv driver is a mapping layer.  It tells you what pointers map to which underlying memory.  Copying that elsewhere in preparation for remapping or reuse or whatever is part of the "what you do with that memory" layer, and AFAICS shouldn't really overlap?

Basically why can't this be a separate driver that just uses mm_drv to do its work?  Or a separate layer within the app memory management code?  I guess I don't see why we need to add new APIs for "weird' hardware to what otherwise looks like a very portable MMU interface.

**Comment by marcinszkudlinski** [comment]
@andyross yes and no. 
The code is to  be called at the very last moment before the device is switched off. All devices - including memory management, should be turned off at the moment. Even worse when restoring the memory - nothing is started yet. So I want to avoid any unnecessary calls, dependencies etc. and put the save/restore routines in the tlb driver.

Another reason - I really don't care for virtual memory mapping when saving the memory contents - all I need is save all used *physical* memory banks. The knowledge about usage of physical banks is locked in a TLB driver and only there.

File: soc/xtensa/intel_adsp/ace_v1x/power.c [hunk 8]
```diff
@@ -5,6 +5,9 @@
  */
 #include <zephyr/zephyr.h>
 #include <zephyr/pm/pm.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+#include <device.h>
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace_v1x/power.c:10]: zephyr/device.h ?

File: soc/xtensa/intel_adsp/ace_v1x/power.c [hunk 9]
```diff
@@ -43,7 +46,18 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
...
+			const struct device *dev = device_get_binding("HPSRAM_TLB");
 
+			if (dev != NULL) {
+				/* the TLB driver suspen will save the whole memory
+				 * leaving the system not operational
+				 * it must be called directly here,
+				 * just before ace_power_down
+				 */
+				pm_device_action_run(dev, PM_DEVICE_ACTION_FORCE_SUSPEND);
+			}
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace_v1x/power.c:59]: This is not correct. If you want to suspend / resume devices manually you have to use device runtime API.

Now the most important thing, devices are automatically suspended / resumed by the power subsystem if they are not busy or doing / using device runtime pm.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 10]
```diff
@@ -20,7 +20,7 @@
  * are untouched.
  */
 
-#define DT_DRV_COMPAT intel_adsp_tlb
+#define DT_DRV_COMPAT DT_DRV_COMPAT_ADSP_TLB
```
Comment (ceolin) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:23]: why this change ?

**Review by ceolin — CHANGES_REQUESTED** [review]
The pm flow is not correct.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c (hunk 10 above)
Comment (marcinszkudlinski) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:23]: because the string name "intel_adsp_tlb" is used twice, in .h and in .c files, so better to keep the string in a macro 
If you disagree I'll change it


mm_drv_intel_adsp_mtl_tlb.h:
/* device id from the device tree */
#define DT_DRV_COMPAT_ADSP_TLB intel_adsp_tlb

/* size of TLB table */
#define TLB_SIZE DT_REG_SIZE_BY_IDX(DT_INST(0, **DT_DRV_COMPAT_ADSP_TLB**), 0)


mm_drv_intel_adsp_mtl_tlb.c:
#define DT_DRV_COMPAT **DT_DRV_COMPAT_ADSP_TLB**

File: soc/xtensa/intel_adsp/ace_v1x/power.c (hunk 9 above)
Comment (marcinszkudlinski) [reply soc/xtensa/intel_adsp/ace_v1x/power.c:59]: I was under impression that Zephyr won't call it automatically unless it is enabled for a device:

bool pm_device_runtime_is_enabled(const struct device *dev)
{
	struct pm_device *pm = dev->pm;

	return pm && atomic_test_bit(&pm->flags, PM_DEVICE_FLAG_RUNTIME_ENABLED);
}

Can't let Zephyr do it automatically, it is too specific/sensitive and must be called right before "turning the light off"....
Changing it to specific API

Comment (ceolin) [reply soc/xtensa/intel_adsp/ace_v1x/power.c:59]: so, you have to use device_runtime_pm for this device as I'd suggested. `pm_device_action_run` is low level function, if you are manually suspend/resuming a device you should use device runtime -> https://docs.zephyrproject.org/latest/services/pm/device_runtime.html

File: soc/xtensa/intel_adsp/ace_v1x/power.c (hunk 7 above)
Comment (marcinszkudlinski) [reply soc/xtensa/intel_adsp/ace_v1x/power.c:57]: obsolete, no more PM interface

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi (hunk 2 above)
Comment (ceolin) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:132]: @marcinszkudlinski  the project is removing the label property. Just remove it. Look the comment below about how to get the device handle without it.

File: soc/xtensa/intel_adsp/ace_v1x/power.c (hunk 7 above)
Comment (ceolin) [reply soc/xtensa/intel_adsp/ace_v1x/power.c:57]: `DEVICE_DT_GET(DT_NODELABEL(tlb))`

**Review by ceolin — DISMISSED** [review]
minor issue with the "label" property to be fixed. Other than that +1

**Comment by dcpleung** [comment]
Please rebase first.

File: include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h [hunk 11]
```diff
@@ -1,21 +1,40 @@
...
+/* size of TLB table */
+#define TLB_SIZE DT_REG_SIZE_BY_IDX(DT_INST(0, DT_DRV_COMPAT_ADSP_TLB), 0)
+
+/* numer of L2 memory pages */
+#define L2_SRAM_PAGES_NUM (L2_SRAM_SIZE / CONFIG_MM_DRV_PAGE_SIZE)
+
+/* maximum size of buffer required for memory save */
+#define TLB_MAX_MEMORY_SAVE_BUFFER_SIZE \
+	(L2_SRAM_SIZE + TLB_SIZE + L2_SRAM_PAGES_NUM * sizeof(void *))
+
```
Comment (edersondisouza) [inline include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h:24]: I wonder if all those defines should actually belong to an "internal" `drivers/mm/mm_drv_intel_adsp_mtl_tlb.h` and only the API bits here.

File: include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+#define ZEPHYR_INCLUDE_DRIVERS_INTEL_ADSP_MTL_TLB
+
+/* device id from the device tree */
+#define DT_DRV_COMPAT_ADSP_TLB intel_adsp_tlb
+
+/* size of TLB table */
+#define TLB_SIZE DT_REG_SIZE_BY_IDX(DT_INST(0, DT_DRV_COMPAT_ADSP_TLB), 0)
+
+/* numer of L2 memory pages */
+#define L2_SRAM_PAGES_NUM (L2_SRAM_SIZE / CONFIG_MM_DRV_PAGE_SIZE)
```
Comment (edersondisouza) [inline include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h]: Not sure if adding unused - in this patch - definitions helps with clean-up. Maybe defer adding those to when they are needed?

**Review by edersondisouza — DISMISSED** [review]
Is this PR still relevant? It hasn't been updated for a while...
Edit: Err, just noticed the last comment from 7 days ago - guess I needed to refresh my tab first =D

**Comment by kv2019i** [comment]
@marcinszkudlinski working on an update. Changing status to draft

File: include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h (hunk 12 above)
Comment (marcinszkudlinski) [reply include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h]: changed

**Comment by marcinszkudlinski** [comment]
D3 power flow for ACE soc added. Please review again

File: soc/xtensa/intel_adsp/ace/power.c [hunk 13]
```diff
@@ -218,6 +220,7 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 			imr_layout->imr_state.header.imr_restore_vector =
 					(void *)rom_entry_d3_restore;
 			imr_layout->imr_state.header.imr_ram_storage = global_imr_ram_storage;
+
```
Comment (edersondisouza) [inline soc/xtensa/intel_adsp/ace/power.c:223]: Unrelated to this patch - maybe belong to previous one?

**Review by dcpleung — DISMISSED** [review]

**Review by aborisovich — DISMISSED** [review]

**Review by tmleman — COMMENTED** [review]
LGTM 👍, can't approve since it's my PR

**Review by softwarecki — DISMISSED** [review]

**Comment by marcinszkudlinski** [comment]
Last (I hope...) version. 
Changes since last review: 
 - compilation fixes 
 - cleaner and little bit updates definitions in adsp_memory.h

**Review by dcpleung — DISMISSED** [review]

**Review by softwarecki — DISMISSED** [review]

**Review by jxstelter — DISMISSED** [review]

**Review by aborisovich — DISMISSED** [review]
looks ready to me, lets merge

**Review by edersondisouza — DISMISSED** [review]

**Comment by stephanosio** [comment]
@ceolin can you please take a look?

File: soc/xtensa/intel_adsp/ace/power.c [hunk 14]
```diff
@@ -148,21 +182,78 @@ __asm__(".align 4\n\t"
...
+			/* save HPSRAM - a multi step procedure, executed by a TLB driver
+			 * the TLB driver will change memory mapping
+			 * leaving the system not operational
+			 * it must be called directly here,
+			 * just before power_down
+			 */
+			const struct device *tlb_dev = DEVICE_DT_GET(DT_NODELABEL(tlb));
 
-			power_down(true, uncache_to_cache(&hpsram_mask[0]),
+			__ASSERT_NO_MSG(dev != NULL);
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:246]: Shouldn't it be `tlb_device` ?

**Comment by ceolin** [comment]
Minor comment, other than that looks good.

File: soc/xtensa/intel_adsp/ace/power.c (hunk 14 above)
Comment (marcinszkudlinski) [reply soc/xtensa/intel_adsp/ace/power.c:246]: True! But that means - __ASSERT_NO_MSG macro is off and does not check anything

**Comment by marcinszkudlinski** [comment]
Done, here we go again with (re) approvals...

File: soc/xtensa/intel_adsp/ace/power.c (hunk 14 above)
Comment (kv2019i) [reply soc/xtensa/intel_adsp/ace/power.c]: @marcinszkudlinski Ack, work in progress to make asserts enabled in at least subset of our CI -> https://github.com/thesofproject/sof/pull/6530 -- first attempt failed, but will wbe looking at this again.

**Review by softwarecki — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Review by marcinszkudlinski — DISMISSED** [review]

**Comment by marcinszkudlinski** [comment]
The PR was approved previously couple of times. Please go ahead and merge it

**Review by aborisovich — DISMISSED** [review]
Just do it

**Review by dcpleung — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by marc-hb** [comment]
commit c929bbcc58c545 fails like this:

```
sof/src/platform/intel/ace/lib/pm_runtime.c
In file included from zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm/ace_v1x-regs.h:11,
                 from sof/src/platform/intel/ace/lib/pm_runtime.c:12:
zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm/adsp_memory.h:43: error: "IMR_BOOT_LDR_MANIFEST_BASE" redefined [-Werror]
   43 | #define IMR_BOOT_LDR_MANIFEST_BASE      (L3_MEM_BASE_ADDR + IMR_BOOT_LDR_MANIFEST_OFFSET)
      | 
In file included from sof/zephyr/../src/../xtos/include/sof/lib/memory.h:11,
                 from sof/zephyr/../src/include/sof/sof.h:13,
                 from sof/zephyr/../src/../xtos/include/sof/lib/pm_runtime.h:19,
                 from sof/src/platform/intel/ace/lib/pm_runtime.c:5:
sof/zephyr/../src/platform/meteorlake/include/platform/lib/memory.h:52: note: this is the location of the previous definition
   52 | #define IMR_BOOT_LDR_MANIFEST_BASE      0xA1042000
      | 
cc1: all warnings being treated as errors
```

Commit 84c09433b574 fails like this:

```
In file included from zephyr/soc/xtensa/intel_adsp/common/include/cpu_init.h:9,
                 from zephyr/soc/xtensa/intel_adsp/ace/power.c:9:
zephyr/soc/xtensa/intel_adsp/ace/power.c: In function 'pm_state_set':
zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm/adsp_memory.h:18:22: error: passing argument 2 of 'memcpy' makes pointer from integer without a cast [-Werror=int-conversion]
   18 | #define LP_SRAM_BASE (DT_REG_ADDR(DT_NODELABEL(sram1)))
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                      |
      |                      long long int
zephyr/soc/xtensa/intel_adsp/ace/power.c:236:56: note: in expansion of macro 'LP_SRAM_BASE'
  236 |                         memcpy(global_imr_ram_storage, LP_SRAM_BASE, LP_SRAM_SIZE);
      |                                                        ^~~~~~~~~~~~
In file included from zephyr/soc/xtensa/intel_adsp/common/include/soc.h:9,
                 from zephyr/include/zephyr/arch/xtensa/irq.h:60,
                 from zephyr/include/zephyr/arch/xtensa/arch.h:27,
                 from zephyr/include/zephyr/arch/cpu.h:27,
                 from zephyr/include/zephyr/kernel_includes.h:33,
                 from zephyr/include/zephyr/kernel.h:17,
                 from zephyr/soc/xtensa/intel_adsp/ace/power.c:6:
zephyr/lib/libc/minimal/include/string.h:42:63: note: expected 'const void * restrict' but argument is of type 'long long int'
   42 | extern void  *memcpy(void *ZRESTRICT d, const void *ZRESTRICT s,
      |                                         ~~~~~~~~~~~~~~~~~~~~~~^
zephyr/soc/xtensa/intel_adsp/ace/power.c: At top level:
zephyr/soc/xtensa/intel_adsp/ace/power.c:168:9: error: '_restore_core_context' is static but used in inline function 'power_off_exit' which is not static [-Werror]
  168 |         _restore_core_context();
      |         ^~~~~~~~~~~~~~~~~~~~~
zephyr/soc/xtensa/intel_adsp/ace/power.c: In function 'pm_state_imr_restore':
zephyr/soc/xtensa/intel_adsp/ace/power.c:198:1: error: 'noreturn' function does return [-Werror]
  198 | }
      | ^
cc1: all warnings being treated as errors

```

**Comment by marc-hb** [comment]
Just found https://github.com/thesofproject/sof/pull/6613, looks like a mutual dependency. Thank @marcinszkudlinski for linking the two.
